node/consensus/src/transactions_queue.rs

```diff
@@ -0,0 +1,354 @@
+            if priority_fee.is_zero() {
+                self.queue.get_or_insert(transaction_id, || transaction);
+            } else {
+                self.priority_queue.insert(transaction_id, transaction, priority_fee);
+            }
+
+            return Ok(());
+        }
+
+        match (self.priority_queue.len() < self.capacity, *priority_fee) {
+            // Invariant: if the queue is at capacity but the priority queue
+            // isn't equal to the capacity, the low-priority queue must be non-empty.
+            (true, 0) => {
+                let _ = self.queue.get_or_insert(transaction_id, || transaction);
+            }
+            (true, _fee) => {
+                // Remove an entry from the low-priority queue to make room for the high-priority transaction.
+                self.queue.pop_lru();
+                self.priority_queue.insert(transaction_id, transaction, priority_fee)
+            }
+
+            // Invariant: if the queue is at capacity but the priority queue is
+            // equal to the capacity, the low-priority queue must be empty.
+            (false, 0) => bail!("The memory pool is full"),
+            (false, _fee) => self.priority_queue.compare_insert(transaction_id, transaction, priority_fee),
+        }
+
+        Ok(())
+    }
+
+    pub fn pop(&mut self) -> Option<(N::TransactionID, Transaction<N>)> {
+        self.priority_queue.pop().or_else(|| self.queue.pop_lru())
+    }
+}
+
+struct PriorityQueue<N: Network> {
+    /// A counter to ensure fifo ordering for transmissions with the same fee.
+    counter: u64,
+    /// A map of transmissions ordered by fee and by fifo sequence.
+    transaction_ids: BTreeMap<(Reverse<U64<N>>, u64), N::TransactionID>,
+    /// A map of transmission IDs to transmissions.
+    transactions: HashMap<N::TransactionID, Transaction<N>>,
+}
+
+impl<N: Network> Default for PriorityQueue<N> {
+    /// Initializes a new instance of the priority queue.
+    fn default() -> Self {
+        Self { counter: Default::default(), transaction_ids: Default::default(), transactions: Default::default() }
+    }
+}
+
+impl<N: Network> PriorityQueue<N> {
+    fn len(&self) -> usize {
+        self.transactions.len()
+    }
+
+    fn insert(&mut self, transaction_id: N::TransactionID, transaction: Transaction<N>, fee: U64<N>) {
+        if let Entry::Vacant(entry) = self.transactions.entry(transaction_id) {
+            // Insert the transaction into the map.
+            entry.insert(transaction);
+            // Sort by fee (highest first) and counter (fifo).
+            self.transaction_ids.insert((Reverse(fee), self.counter), transaction_id);
+            // Increment the counter.
+            self.counter += 1;
+        }
+    }
+
+    fn compare_insert(&mut self, transaction_id: N::TransactionID, transaction: Transaction<N>, fee: U64<N>) {
+        // Make sure the collection isn't empty.
+        if self.transaction_ids.is_empty() {
+            return;
```

Comment:
Same as the other case. Why not error here? The TX was not actually added to the queue.